swf/src/types.rs

```diff
@@ -97,7 +97,7 @@ pub struct HeaderExt {
     pub(crate) header: Header,
     pub(crate) file_attributes: FileAttributes,
     pub(crate) background_color: Option<SetBackgroundColor>,
-    pub(crate) uncompressed_len: u32,
+    pub(crate) uncompressed_len: i32,
```

Comment:
I feel a little weird letting negative lengths leak into the SWF parsing machinery. Did Flash have a hard 2GB cap on SWF size (even on 64-bit)?

Reply:
I understand that. Sadly we can't just use an unsigned integer here because `-1` is a valid value needed in some situations (and `0` is as well, so we can't just read `0` as `-1`).
One alternative to this I easily see would be an `Option<u32>` (with `None` standing for `-1`). However, this would make comparisons to other integers much harder since these multiple cases would need to be handled in each comparison (or a helper function would need to be used, which also isn't that clean).

The other possible alternative to this I see would be a custom `enum` with `u32` and a dedicated `MinusOne` value. This has the advantage that `PartialEq<u32>` and `PartialOrd<u32>` could be implemented for it (and vice versa), so that all comparisons could just work without changing them in any way. This might make sense in general because there are other variables in Ruffle that can just be positive or `-1` as well (e.g. frames_loaded which this pull request has also changed (from `u16` to `i32`)).

What do you think about those ideas?

And no, at least I'm not aware of any hard SWF file size cap. I just used `i32` because it has previously been `u32` and the 2GB are so large that it seemed to make more sense to change it to `u32` than `u64`. The same thought applies to a 4GB file with all the `u32`s used to store the SWF file size though.

Reply:
Would it be possible to store the "this is actually an error SWF" flag in `SwfMovie`, and then have `SwfMovie.uncompressed_len()` check for that and return `-1` as necessary? (And, of course, not access `SwfMovie.header().uncompressed_len` directly)

I'm less concerned about the file size cap and more with letting details of how Ruffle works determine `swf`'s interface. If you're doing low-level SWF stuff with `swf` you only care about the uncompressed file size and not any error conditions that might have been signaled by Flash Player abusing that field.

Reply:
That might be a possibility as well. Although this error condition can only appear if there literally is no existing SWF file, so in any case where it would be `-1`, there is no actual file size you could care about.

Still, if you prefer it, I can change it to this. The only thing worrying me about this is that the identically named functions `SwfMovie::uncompressed_len` and `HeaderExt::uncompressed_len` would not always return the same value (although this would be mitigated by documentation). And there are some places where `HeaderExt::uncompressed_len` is directly used: To create `MovieMetadata` in `web/src/lib.rs` and to set the `FileResults` in `scanner/src/execute.rs`. I don't know if these values are furtherly used elsewhere and problems arise if these values used directly from the header diverge from the `-1` error values retrieved by the `MovieClip`.
I could change these places to use the `SwfMovie::uncompressed_len` value, but I'm not sure if it would be good to do that or if there are reasons why they take it from the header directly and maybe should not be `-1` in such an error situation.
What would you think about that all?

Reply:
I've played around with this for a while and reread your previous statement, and I noticed your "(And, of course, not access `SwfMovie.header().uncompressed_len` directly)", which I think would probably mean that I should change `web/src/lib.rs` and `scanner/src/execute.rs` to use `SwfMovie::uncompressed_len` instead.
Since my biggest worry was that there were multiple variables containing different values with the same name (`uncompressed_len`) and that both of these are used in many cases, resulting in two diverging values being used, this would be mitigated by just using `SwfMovie::uncompressed_len` everywhere (and adding documentation stating that only it should be used).

I still don't really understand what speaks against just directly setting `HeaderExt::uncompressed_len` to the real Flash value. I personally think that using something like an `U32OrMinusOne` enum would be a good solution.
In this error case, no SWF file and no "correct" value exists (as there is no file size); the whole header is then just consisting of the error values Flash uses, so in my opinion, having a dedicated error value would be coherent and made sense in the header. An `U32OrMinusOne` enum allows the correct range of possible values.
And by using this, we could be sure `uncompressed_len` is always the same and that there aren't diverging variables with the same name, no matter which function is used (or will be added in the future) to get it, so for me, this look a cleaner solution.

However, these are just my thoughts about it. If there are other reasons why you prefer it with a flag, that's fine with me as well. We could also still use such an enum as the `HeaderExt::uncompressed_len` return value; that would solve the 2GB cap and be a clean way to allow `-1` but not other illegal negative values. The only downside to that enum I see is that it needs 8 bytes (as much as an `i64`).

Reply:
Y'know, I think I might be overthinking this. Let's just keep this as-is for now. If it causes a problem later on we can fix it.